### PR TITLE
Adapt Robin / Neumann boundary conditions to user needs.

### PR DESCRIPTION
**Pull request by TomFischer**
Add possibility to specify 'lower' dimensional Robin/Neumann boundary conditions.

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.2.2)
2. [x] Tests covering your feature were added?
3. [x] Any new feature or behavior change was documented?

### CODE REVIEW

**Comment by TomFischer** [comment]
Rebased and changed tolerances in ctests.

**Comment by endJunction** [comment]
Please review.

**Comment by TomFischer** [comment]
Rebased again. Please a second reviewer!

File: ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h [hunk 1]
```diff
@@ -64,17 +65,29 @@ class RobinBoundaryConditionLocalAssembler final
...
         {
+            position.setIntegrationPoint(ip);
             auto const& ip_data = Base::_ns_and_weights[ip];
             auto const& N = ip_data.N;
             auto const& w = ip_data.weight;
 
+            double integral_measure = 1.0;
+            if (_data.integral_measure)
+            {
+                integral_measure = (*_data.integral_measure)(t, position)[0];
```
Comment (wenqing) [inline ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h]:
```suggestion
                const double integral_measure =   (_data.integral_measure) ? (*_data.integral_measure)(t, position)[0] : 1.0;
```

File: ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h [hunk 2]
```diff
@@ -58,24 +64,36 @@ class NeumannBoundaryConditionLocalAssembler final
...
             auto const& ip_data = Base::_ns_and_weights[ip];
+            auto const& N = ip_data.N;
+            auto const& w = ip_data.weight;
 
-            _local_rhs.noalias() += ip_data.N *
-                                    parameter_node_values.dot(ip_data.N) *
-                                    ip_data.weight;
+            if (_data.integral_measure)
+            {
+                integral_measure = (*_data.integral_measure)(t, position)[0];
```
Comment (wenqing) [inline ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h]:
```suggestion
               const double  integral_measure = (_data.integral_measure) ? (*_data.integral_measure)(t, position)[0] : 1.0;
```

File: ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h [hunk 3]
```diff
@@ -58,24 +64,36 @@ class NeumannBoundaryConditionLocalAssembler final
...
         NodalVectorType parameter_node_values =
-            _neumann_bc_parameter.getNodalValuesOnElement(Base::_element, t)
+            _data.neumann_bc_parameter
+                .getNodalValuesOnElement(Base::_element, t)
                 .template topRows<ShapeFunction::MeshElement::n_all_nodes>();
 
+        ParameterLib::SpatialPosition position;
+        position.setElementID(Base::_element.getID());
+
+        double integral_measure = 1.0;
```
Comment (wenqing) [inline ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h]: remove it if the following suggestion (fro lines 82-84) is taken .

File: ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h (hunk 1 above)
Comment (TomFischer) [reply ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h]: I can't declare/define the variable `integral_measure` only in the scope of the branch of the if condition since it is used below.

File: ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h (hunk 2 above)
Comment (TomFischer) [reply ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h]: I think this doesn't work. See the argument in the RobinBoundaryConditionLocalAssembler case.

Comment (wenqing) [reply ProcessLib/BoundaryCondition/NeumannBoundaryConditionLocalAssembler.h]: I see.

File: ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h (hunk 1 above)
Comment (wenqing) [reply ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h]: I see.

Comment (wenqing) [reply ProcessLib/BoundaryCondition/RobinBoundaryConditionLocalAssembler.h]: :fast_forward:

**Comment by endJunction** [comment]
@TomFischer Something is consistently wrong with the parallel tests:
```
[2019-11-14T13:22:03.107Z] 	  2 - ogs-Parallel_2D_StaggeredScheme_ComponentTransport_TracerSimulation-mpirun (Timeout)
[2019-11-14T13:22:03.107Z] 	  3 - ogs-Parallel_2D_StaggeredScheme_ComponentTransport_TracerSimulation-mpirun-vtkdiff (Failed)
[2019-11-14T13:22:03.107Z] 	  8 - ogs-ParallelFEM_GroundWaterFlow3D_NeumannBC-mpirun (Timeout)
[2019-11-14T13:22:03.107Z] 	  9 - ogs-ParallelFEM_GroundWaterFlow3D_NeumannBC-mpirun-vtkdiff (Failed)
```
I restarted the parallel jenkins job yesterday with same result...

**Comment by endJunction** [comment]
I'd think that the `integral_measure` should not be 1 in an axisymmetric setup. A test would be nice... Let's discuss it on Monday.
